core/include/vecmem/containers/details/jagged_vector_view.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "vecmem/containers/details/vector_view.hpp"
+
+#include <cstddef>
+
+namespace vecmem::details {
+    /**
+     * @brief A view for jagged vectors.
+     *
+     * A jagged vector is a two-dimensional vector in which the inner vectors do
+     * not necessarily have the same size. For example, a jagged vector might
+     * look like this:
+     *
+     * [[0, 1, 2],
+     *  [3, 4],
+     *  [],
+     *  [5, 6, 7]]
+     *
+     * This class is a view of existing two-dimensional vectors created using a
+     * vector-of-vectors formalism. Elements cannot be added or removed through
+     * this view, but individual elements can be accessed and modified.
+     *
+     * @warning This view class shares memory with the vectors from which it was
+     * constructed. Operating on the underlying vectors while an instance of
+     * this class exists deriving from it is undefined and may leave the view in
+     * an undefined state.
+     */
+    template<typename T>
+    struct jagged_vector_view {
+        jagged_vector_view(
+            std::size_t size,
+            vector_view<T> * ptr
+        );
+
+        /**
+         * The number of rows in this jagged vector.
+         */
+        const std::size_t m_size;
+
+        /**
+         * The internal state of this jagged vector, which is heap-allocated by
+         * the given memory manager.
+         */
+        vector_view<T> * const m_ptr;
```

Comment:
I don't really agree with this argument. :frowning: Having a view point to invalid memory is really dangerous since it could lead to unrecoverable segmentation faults. Allowing the user to set the pointers and size manually invites that kind of behaviour, while constructing them as const and having our code set up these values allows us to enforce more invariants. In what case would it be warranted for the user to change what a view points at, rather than just creating a new view? These objects are super simple and cheap to construct!

Reply:
Having written some amount of template meta-code in the past, dealing with types that have no default constructors, and/or no assignment operators, can be a real chore. It's in such complicated templating that I can very well imagine a non-valid object being constructed, which would get valid member values set up on it after some initialisation. Yes, such things can be avoided if necessary, but many times only at great (coding) cost.

In general, this really mostly comes from experience. Every time that I declared a variable as const in any of my classes, eventually I always had to remove those declarations. As sooner or later I ran into something that would've necessitated it. (This comes "from the same place" as why I never use reference member values in my classes, only [smart] pointers.)

Reply:
That's fair, I'll trust your judgement on it then!